### PR TITLE
[Segment Replication] Add update doc integ test for Segment replication

### PR DESCRIPTION
**Pull request by dreamer-89**
Signed-off-by: Suraj Singh <surajrider@gmail.com>

### Description
This PR adds update document integration tests in order to verify the fix (https://github.com/opensearch-project/OpenSearch/pull/4185) works for update operations. 

Verified without fix above, the added update IT fails with below error
```
[2022-08-16T20:03:03,300][ERROR][o.o.i.r.SegmentReplicationTargetService] [node_t1] replication failure
org.opensearch.OpenSearchException: Segment Replication failed
	at org.opensearch.indices.replication.SegmentReplicationTargetService$3.onFailure(SegmentReplicationTargetService.java:235) [main/:?]
...
Caused by: java.lang.IllegalStateException: Shard [test-idx-1][0] has local copies of segments that differ from the primary
	at org.opensearch.indices.replication.SegmentReplicationTarget.getFiles(SegmentReplicationTarget.java:167) ~[main/:?]
	... 24 more
```
 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/3787

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java [hunk 1]
```diff
@@ -299,6 +303,60 @@ public void testDeleteOperations() throws Exception {
         }
     }
 
+    public void testUpdateOperations() throws Exception {
+        final String primary = internalCluster().startNode();
+        createIndex(INDEX_NAME);
+        ensureYellow(INDEX_NAME);
```
Comment (Rishikesh1159) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:309]: Just for my understanding, why aren't we doing ensureGreen() after starting both primary and replica as we do for rest of tests in this class?

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java [hunk 2]
```diff
@@ -299,6 +303,60 @@ public void testDeleteOperations() throws Exception {
...
+            waitForReplicaUpdate();
+
+            assertHitCount(client(primary).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), expectedHitCount);
+            assertHitCount(client(replica).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), expectedHitCount);
+
+            Set<String> ids = indexer.getIds();
+            String id = ids.toArray()[0].toString();
+            UpdateResponse updateResponse = client(primary).prepareUpdate(INDEX_NAME, id)
+                .setDoc(Requests.INDEX_CONTENT_TYPE, "foo", "baz")
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.WAIT_UNTIL)
```
Comment (Rishikesh1159) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:346]: Same here for my understanding, why are we doing "WAIT_UNTIL" but not "IMMEDIATE"

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java (hunk 1 above)
Comment (dreamer-89) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:309]: Starting with one node ensures that node is the primary. In other tests, any node can hold the primary shard.

Comment (Rishikesh1159) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:309]: Makes sense. Thank you!

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java (hunk 2 above)
Comment (dreamer-89) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:346]: `Wait_until` needs explicit `refresh` before making document searchable, while `IMMEDIATE` makes explicity shard refresh calls to make them searchable as part of operation. You can check other refresh policies in `RefreshPolicy` enum under WriteRequest.java

Comment (nknize) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:346]: Agree w/ this approach...

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java [hunk 3]
```diff
@@ -41,8 +43,10 @@
...
 import java.util.stream.Collectors;
 
-import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.index.query.QueryBuilders.matchQuery;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertSearchHits;
 
 @OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 0)
 public class SegmentReplicationIT extends OpenSearchIntegTestCase {
```
Comment (nknize) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:52]: Could we just make this extend `OpenSearchSingleNodeTestCase`?

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java (hunk 2 above)
Comment (Rishikesh1159) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:346]: Makes sense. Thank you !

**Review by Rishikesh1159 — APPROVED** [review]

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java (hunk 3 above)
Comment (dreamer-89) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:52]: Thanks @nknize for the review. There are certain tests in this class which needs multiple nodes (specially failover related), so I think we can not use `OpenSearchSingleNodeTestCase` class. Please correct me if I am missing something.

**Pull request merged**